sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -579,6 +573,14 @@ case class DataSource(
     DataSource.checkAndGlobPathIfNecessary(allPaths.toSeq, newHadoopConfiguration(),
       checkEmptyGlobPath, checkFilesExist, enableGlobbing = globPaths)
   }
+
+  private def checkAllowedTypesInWrite(dataTypes: Seq[DataType]): Unit = {
+    dataTypes.foreach {
+      case i @ (CalendarIntervalType | DayTimeIntervalType | YearMonthIntervalType) =>
```

Comment:
we should use `DataType.exists` to detect nested interval types.

Reply:
At least, this is compatible with current master which checks intervals **only** on top-level. :-)  See
On top-level:
```
scala> spark.range(1).selectExpr("""timestamp'2021-01-02 00:01:02' - timestamp'2021-01-01 00:00:00'""").write.mode("overwrite").parquet("/Users/maximgekk/tmp/123")
org.apache.spark.sql.AnalysisException: Cannot save interval data type into external storage.
```

Nested:
```
scala> spark.range(1).selectExpr("""struct(timestamp'2021-01-02 00:01:02' - timestamp'2021-01-01 00:00:00')""").write.mode("overwrite").parquet("/Users/maximgekk/tmp/123")
org.apache.spark.sql.AnalysisException: Parquet data source does not support struct<col1:interval> data type.
```
The last exception come from another place. Parquet cannot save nested interval because `CalendarIntervalType` is not `AtomicType`.

Reply:
There also a method here`org.apache.spark.sql.catalyst.util.TypeUtils.failWithIntervalType`, maybe we can merge them together

Reply:
since we are touching it, let's make it right and consider nested interval types as well.